File: hud/view/device/select_device_panel.py
import asyncio
from typing import Union

from PySide6.QtCore import Qt
from PySide6.QtWidgets import QMainWindow, QGridLayout, QWidget, QSpacerItem, QSizePolicy

from hud.configuration.config import Config
from hud.model import Service
from hud.model.data_classes import Device
from hud.model.model import Model
from hud.view import DeviceController
from hud.view.device_dialog import DeviceDialog
from hud.view.primitives.labels import ClickableLabel, LabelType, ValueLabel, LabelSize
from hud.view.primitives.theme_switch import with_switchable_theme


@with_switchable_theme
class SelectDevicePanel(QMainWindow):

    def __init__(
            self,
            normal_icon_path: str,
            highlighted_icon_path: str,
            controller: DeviceController,
            ble_service_type: Service,
            model: Model,
            app_config: Config,
            parent=None
    ):
        super().__init__(parent)
        self.app_config = app_config
        self.dialog = None

        self.model = model
        self.ble_service_type = ble_service_type
        self.controller = controller
        self.highlighted_icon_path = highlighted_icon_path
        self.normal_icon_path = normal_icon_path

        self.selectIcon = ClickableLabel(
            normal_icon_path=self.normal_icon_path,
            highlighted_icon_path=self.highlighted_icon_path,
            theme=self.app_config.hud_layout.theme,
            label_size=LabelType.CLICKABLE_LABEL,
        )
        self.selectIcon.setToolTip("No device selected")
        self.selectIcon.clicked.connect(lambda: asyncio.ensure_future(self.showSelectDeviceDialog()))

        self.metricLabel = ValueLabel("--/--", self.app_config.hud_layout.theme, size=LabelSize.SMALL)
        self.metricLabel.setToolTip("No device selected")

        self.layout = QGridLayout()
        self.layout.addWidget(self.selectIcon, 0, 0, 1, 1, Qt.AlignmentFlag.AlignCenter)
        self.layout.addItem(
            QSpacerItem(20, 10, QSizePolicy.Policy.Minimum, QSizePolicy.Policy.Expanding), 1, 0, 1, 1,
            Qt.AlignmentFlag.AlignCenter
        )
        self.layout.addWidget(self.metricLabel, 2, 0, 1, 1, Qt.AlignmentFlag.AlignCenter)

        self.centralWidget = QWidget(self)
        self.centralWidget.setLayout(self.layout)
        self.setCentralWidget(self.centralWidget)

    async def showSelectDeviceDialog(self):
        self.dialog = DeviceDialog(self.app_config, self)
        self.dialog.selectDeviceSignal.connect(lambda x: asyncio.ensure_future(self.deviceSelected(x)))
        self.populate_device_list(self.model.find_devices(self.ble_service_type))

        asyncio.ensure_future(self.controller.start_device_scan())
        asyncio.ensure_future(self.dialog.show_async())

    def populate_device_list(self, devices: Union[Device, list[Device]]):
        if not self.dialog:
            return

        data = [devices] if isinstance(devices, Device) else devices

        for device in data:
            self.dialog.showDevice(device)

    async def deviceSelected(self, device: Device):
        self.dialog = None

        if device is None:
            return

        asyncio.ensure_future(self.controller.set_device(device))

    def updateDevice(self, value):
        self.selectIcon.setToolTip(str(value))
        self.metricLabel.setToolTip(str(value))

    def updateMetrics(self, value):
        data = str(value.latest) if hasattr(value, "latest") else "ok"
        self.metricLabel.setText(data)

    def closeEvent(self, event):
        event.accept()

    def bind_to_model(self, channel):
        channel.device_selected.subscribe(self.updateDevice)
        channel.metrics.subscribe(self.updateMetrics)
        channel.device_discovered.subscribe(self.populate_device_list)

File: src/SelectDevicePanel.cpp
#include <QMainWindow>
#include <QGridLayout>
#include <QWidget>
#include <QSpacerItem>
#include <QSizePolicy>

#include "SelectDevicePanel.h"

#include <iostream>

#include "StyleSheets.h"

#include <iostream>
#include <qevent.h>

#include "Constants.h"
#include "ControllerHandler.h"
#include "ModelEvents.h"
#include "SupportedServices.h"

SelectDevicePanel::SelectDevicePanel(
    const AppService &service,
    const std::string &normal_icon_path,
    const std::string &highlighted_icon_path,
    const std::shared_ptr<ControllerHandler> &handler,
    QWidget *parent
) : QMainWindow(parent), handler(handler), service(service) {
    selectIcon = new ClickableLabel(normal_icon_path, highlighted_icon_path, this);
    selectIcon->setToolTip("No device selected");
    connect(selectIcon, &ClickableLabel::clicked, this, &SelectDevicePanel::handleDeviceButtonClick);
    metricLabel = new ValueLabel("--/--", LabelSize::MEDIUM, this);
    metricLabel->setToolTip("No device selected");

    const auto spacer = new QSpacerItem(20, 20, QSizePolicy::Minimum, QSizePolicy::Expanding);

    const auto layout = new QGridLayout(this);
    layout->addWidget(selectIcon, 0, 0, Qt::AlignCenter);
    layout->addItem(spacer, 1, 0, Qt::AlignCenter);
    layout->addWidget(metricLabel, 2, 0, Qt::AlignCenter);

    const auto centralWidget = new QWidget(this);
    centralWidget->setLayout(layout);
    setCentralWidget(centralWidget);

    setStyleSheet((StyleSheets::THEME_DARK + StyleSheets::SCALE_MEDIUM).data());
}

void SelectDevicePanel::deviceSelected(const DeviceSelected &event) {
    if (event.service != this->service.service) {
        return;
    }

    std::cout << "SelectDevicePanel::deviceSelected" << std::endl;
    const auto name = QString::fromStdString(event.device->name.value);
    if (selectIcon) {
        selectIcon->setToolTip(name);
    }
    if (metricLabel) {
        metricLabel->setToolTip(name);
    }
    setToolTip(name);
}

void SelectDevicePanel::measurementsReceived(const WorkoutData &measurements_update) {
    std::cout << "SelectDevicePanel::measurementsReceived" << std::endl;

    if (service.service == HEART_RATE.service) {
        std::cout << "  SelectDevicePanel::measurementsReceived: HEART_RATE" << std::endl;
        const auto heart_rate = measurements_update.hrm.val;
        metricLabel->setText(QString::number(heart_rate));
    }

    if (service.service == CADENCE.service) {
        std::cout << "  SelectDevicePanel::measurementsReceived: CADENCE" << std::endl;
        const auto cadence = measurements_update.cadence.val;
        metricLabel->setText(QString::number(cadence));
    }

    if (service.service == SPEED.service) {
        std::cout << "  SelectDevicePanel::measurementsReceived: SPEED" << std::endl;
        const auto speed = measurements_update.speed.val;
        metricLabel->setText(QString::number(speed));
    }

    if (service.service == POWER.service) {
        std::cout << "  SelectDevicePanel::measurementsReceived: POWER" << std::endl;
        const auto power = measurements_update.power.val;
        metricLabel->setText(QString::number(power));
    }
}

void SelectDevicePanel::handleDeviceButtonClick() const {
    std::cout << "SelectDevicePanel::show_device_dialog()" << std::endl;
    handler->next(Constants::Screens::DEVICE_DIALOG);
}
